spinnaker-monitoring-daemon/spinnaker-monitoring/stackdriver_handlers.py

```diff
@@ -207,18 +225,360 @@ def descriptors_to_text(self, descriptor_list):
+      return '%d %s' % (rank, info.type_name)
+
+    return sorted(survey.values(), key=sort_key)
+
+  def build_survey_from_type_names(self, type_names):
+    """Returns a dict of SurveyInfo keyed by descriptor type_name.
+
+    Args:
+      type_names: [list of string]
+    """
+    num_threads = len(self.__op_methods)
+
+    pool = ThreadPool(num_threads)
+    process = lambda type_name: self.__survey_type(type_name)
+    survey = pool.map(process, type_names)
+    pool.close()
+    pool.join()
+    return {info.type_name: info for info in survey}
+
+  def __survey_type(self, type_name):
+    """Performs survey on a given descriptor type name.
+
+    Returns:
+      SurveyInfo
+    """
+    method = self.__op_methods.pop()
+    op = method(filter='metric.type="%s"' % type_name, **self.__op_params)
+    try:
+      # Stackdriver likes to return random 500 errors,
+      # especially when clients hammer it, so retry.
+      attempt = 1
+      while True:
+        info, can_retry = self.__attempt_op(op, type_name)
+        if not info.error or not can_retry:
+          return info
+        if attempt == self.__max_attempts:
+          return info
+
+        logging.error('Retryable error processing %s...', type_name)
+        time.sleep(self.__secs_between_attempts)
+        attempt += 1
+        continue
+    finally:
+      # put back
+      self.__op_methods.append(method)
+
+  def __attempt_op(self, op, type_name):
+    """Return SurveyInfo, can_retry."""
+
+    try:
+      result = op.execute()
+      if not result:
+        return SurveyInfo.make_unknown(type_name), False
+
+      if 'timeSeries' not in result:
+        logging.error('Unexpected result surveying "%s": %s',
+                      type_name, result)
+        return SurveyInfo.make_error(type_name, result), True
+
+      last_iso = result['timeSeries'][0]['points'][0]['interval']['endTime']
+      last_time = datetime.datetime.strptime(last_iso, self.__date_format)
+      days_ago = (self.__today - last_time).days
+      return SurveyInfo.make_seen(type_name, last_iso, days_ago), True
+
+    except HttpError as err:
+      retryable = err.resp.status >= 500
+      logging.error('Error processing %s: %s', type_name, str(err))
+      error = 'HTTP [%d]: %s ' % (err.resp.status, err.resp.reason)
+      return SurveyInfo.make_error(type_name, error), retryable
+
+    except Exception as e:
+      logging.error('Error processing %s: %s', type_name, str(e))
+      return SurveyInfo.make_error(type_name, str(e)), False
+
+
+class AuditCustomDescriptorsHandler(BaseStackdriverCommandHandler):
+  """Administrative handler to distinguish in-use from unused descriptors."""
+
+  def delete_unused(self, options, surveyor, survey):
+    descriptor_name_prefix = 'projects/{project}/metricDescriptors/'.format(
+        project=surveyor.project)
+    to_delete = []
+    for info in survey.values():
+      if info.days_ago == info.UNKNOWN_DAYS_AGO:
```

Comment:
Does `UNKNOWNS_DAYS_AGO` effectively mean that the metric was unused? Are we filtering these down to the metric descriptors Spinnaker uses or might we be affecting descriptors in the project not created by Spinnaker?

Reply:
This is among the metrics (and timeframe) given it. In practice we filter to spinnaker, but I wrote the class so it could be lifted outside spinnaker since this seems like a fundamental usability issue with the API.